src/main/java/io/lighty/yang/validator/simplify/SchemaSelector.java

```diff
@@ -67,44 +68,70 @@ private void fillUsedSchema(final InputStream input, final SchemaTree st)
     }
 
     public void noXml() {
+        SchemaInferenceStack schemaInferenceStack = SchemaInferenceStack.of(effectiveModelContext);
         for (Module module : effectiveModelContext.getModules()) {
             for (DataSchemaNode node : module.getChildNodes()) {
-                resolveChildNodes(tree, node, true, false);
+                resolveChildNodes(tree, node, true, false, schemaInferenceStack);
             }
 
             for (AugmentationSchemaNode aug : module.getAugmentations()) {
+                schemaInferenceStack.enterSchemaTree(aug.getTargetPath());
                 for (DataSchemaNode node : aug.getChildNodes()) {
-                    resolveChildNodes(tree, node, true, true);
+                    resolveChildNodes(tree, node, true, true, schemaInferenceStack);
                 }
+                schemaInferenceStack.clear();
             }
         }
     }
 
     private void resolveChildNodes(final SchemaTree schemaTree, DataSchemaNode node, boolean rootNode,
-                                   boolean augNode) {
-        SchemaTree childSchemaTree = schemaTree.addChild(node, rootNode, augNode);
+                                   boolean augNode, final SchemaInferenceStack schemaInferenceStack) {
+
         if (node instanceof DataNodeContainer) {
+            schemaInferenceStack.enterSchemaTree(node.getQName());
+            SchemaTree childSchemaTree = schemaTree.addChild(node, rootNode, augNode,
+                    schemaInferenceStack.toSchemaNodeIdentifier());
             for (DataSchemaNode schemaNode : ((DataNodeContainer) node).getChildNodes()) {
-                resolveChildNodes(childSchemaTree, schemaNode, false, false);
+                resolveChildNodes(childSchemaTree, schemaNode, false, false, schemaInferenceStack);
             }
+            schemaInferenceStack.exit();
         } else if (node instanceof ChoiceSchemaNode) {
+            schemaInferenceStack.enterChoice(node.getQName());
+            SchemaTree childSchemaTree = schemaTree.addChild(node, rootNode, augNode,
+                    schemaInferenceStack.toSchemaNodeIdentifier());
             final Collection<? extends CaseSchemaNode> cases = ((ChoiceSchemaNode) node).getCases();
             for (DataSchemaNode singelCase : cases) {
-                resolveChildNodes(childSchemaTree, singelCase, false, false);
+                resolveChildNodes(childSchemaTree, singelCase, false, false, schemaInferenceStack);
             }
+            schemaInferenceStack.exit();
+        }
+
+        if (node instanceof DataTreeEffectiveStatement) {
```

Comment:
Why this new if is added?

`if (node instanceof DataTreeEffectiveStatement)`

Reply:
I made this little complicated, I have reworked this. This IF create SchemaTree for elements which were not matched in other If conditions. Before SchemaTree was created on beginning of method. I was thinking that for ChoiceSchemaNode I have to use schemaInferenceStack.enterChoice(node.getQName()), but this is not true.